### PR TITLE
[draft][HIP][CUDA] Use new UR entry point to get native mem with device param

### PR DESCRIPTION
**Pull request by hdelan**
Accessor interop broke with the multi device context for HIP (https://github.com/oneapi-src/unified-runtime/pull/999). This fixes it.

Relates to https://github.com/oneapi-src/unified-runtime/pull/1199

### CODE REVIEW

File: sycl/include/sycl/detail/pi.h [hunk 1]
```diff
@@ -1402,6 +1402,13 @@ __SYCL_EXPORT pi_result piMemBufferPartition(
 __SYCL_EXPORT pi_result piextMemGetNativeHandle(pi_mem mem,
                                                 pi_native_handle *nativeHandle);
 
+/// Gets the native handle of a PI mem object on a particular device
+///
+/// \param mem is the PI mem to get the native handle of.
```
Comment (maarquitos14) [inline sycl/include/sycl/detail/pi.h:1407]: `dev` param description is missing.

File: sycl/test-e2e/HostInteropTask/interop-task-hip.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+template <typename DataT, DataT B1Init, DataT B2Init>
+void init(buffer<DataT, 1> &B1, buffer<DataT, 1> &B2, queue &Q) {
+  Q.submit([&](handler &CGH) {
+    auto Acc1 = B1.template get_access<mode::write>(CGH);
+    auto Acc2 = B2.template get_access<mode::write>(CGH);
+
+    CGH.parallel_for<Init<DataT>>(BUFFER_SIZE, [=](item<1> Id) {
+      Acc1[Id] = -1;
+      Acc2[Id] = -2;
+    });
```
Comment (maarquitos14) [inline sycl/test-e2e/HostInteropTask/interop-task-hip.cpp:82]:
```suggestion
      Acc1[Id] = B1Init;
      Acc2[Id] = B2Init;
    });
```

File: sycl/test-e2e/HostInteropTask/interop-task-hip.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  for (size_t Idx = 0; Idx < COUNT; ++Idx) {
+    copy(Buffer1, Buffer2, Q);
+    modify(Buffer2, Q);
+    copy(Buffer2, Buffer1, Q);
+  }
+  checkBufferValues(Buffer1, COUNT - 1);
+  checkBufferValues(Buffer2, COUNT - 1);
+}
+
+// Same as above but with queue constructed out of context
```
Comment (maarquitos14) [inline sycl/test-e2e/HostInteropTask/interop-task-hip.cpp:131]: I don't see any new queue constructed. Am I missing anything here?

File: sycl/test-e2e/HostInteropTask/interop-task-hip.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+// Same as above, but performing each command group on a separate SYCL queue
+// (on the same or different devices). This ensures the dependency tracking
+// works well but also there is no accidental side effects on other queues.
+void test2(queue &Q) {
+  static constexpr int COUNT = 4;
+  buffer<int, 1> Buffer1{BUFFER_SIZE};
+  buffer<int, 1> Buffer2{BUFFER_SIZE};
+
+  // init the buffer with a'priori invalid data
```
Comment (maarquitos14) [inline sycl/test-e2e/HostInteropTask/interop-task-hip.cpp:118]: Comments should be proper sentences, i.e. start with capital letters and finish with `.`: https://llvm.org/docs/CodingStandards.html#commenting. Please, apply wherever required.

File: sycl/test-e2e/HostInteropTask/interop-task-hip.cpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+int main() {
+  queue Q([](sycl::exception_list ExceptionList) {
+    if (ExceptionList.size() != 1) {
+      std::cerr << "Should be one exception in exception list" << std::endl;
+      std::abort();
+    }
+    std::rethrow_exception(*ExceptionList.begin());
+  });
+  tests(Q);
+  tests(Q);
```
Comment (maarquitos14) [inline sycl/test-e2e/HostInteropTask/interop-task-hip.cpp:196]: Why twice?

Comment (hdelan) [reply sycl/test-e2e/HostInteropTask/interop-task-hip.cpp:196]: Aha I didn't see this. This test was mostly copy pasted from here https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/HostInteropTask/interop-task.cpp which has this testing twice. I don't know why so I can remove this. I am leaving this PR as a draft until people come back from holiday so there is no need to review yet. Thanks

File: sycl/test-e2e/HostInteropTask/interop-task-hip.cpp (hunk 3 above)
Comment (hdelan) [reply sycl/test-e2e/HostInteropTask/interop-task-hip.cpp:131]: Have removed. Thanks

File: sycl/test-e2e/HostInteropTask/interop-task-hip.cpp (hunk 4 above)
Comment (hdelan) [reply sycl/test-e2e/HostInteropTask/interop-task-hip.cpp:118]: Thanks, change made

File: sycl/test-e2e/HostInteropTask/interop-task-hip.cpp (hunk 2 above)
Comment (hdelan) [reply sycl/test-e2e/HostInteropTask/interop-task-hip.cpp:82]: Thanks change made

File: sycl/include/sycl/detail/pi.h (hunk 1 above)
Comment (hdelan) [reply sycl/include/sycl/detail/pi.h:1407]: Have added, thanks.

**Comment by hdelan** [comment]
@steffenlarsen it seems that either the clang format fails or the "Code formatter" thing fails for `sycl/include/sycl/detail.pi.h`. . They seem to have different configs. This is just a draft PR so not important but just wondering if you know anything about this

**Comment by steffenlarsen** [comment]
> @steffenlarsen it seems that either the clang format fails or the "Code formatter" thing fails for `sycl/include/sycl/detail/pi.h`. They seem to have different configs. This is just a draft PR so not important but just wondering if you know anything about this

That is curious! @aelovikov-intel - Do you know if the new formatter check comes from upstream?

**Comment by aelovikov-intel** [comment]
> > @steffenlarsen it seems that either the clang format fails or the "Code formatter" thing fails for `sycl/include/sycl/detail/pi.h`. They seem to have different configs. This is just a draft PR so not important but just wondering if you know anything about this
> 
> That is curious! @aelovikov-intel - Do you know if the new formatter check comes from upstream?

That was intentional - https://github.com/intel/llvm/discussions/12085 / https://github.com/intel/llvm/pull/12184, but not yet complete. I have https://github.com/intel/llvm/pull/12240 in progress and we'll need to disable our old format check job.

**Comment by hdelan** [comment]
Thanks for clarification @steffenlarsen @aelovikov-intel ! This is just a draft for the moment, but will mark as ready for review when I have thumbs up from UR people. Maybe by then the CI linting will be fixed

**Review by maarquitos14 — APPROVED** [review]
LGTM.

**Pull request closed without merging**